src/quv/modules.c

```diff
@@ -114,22 +115,81 @@ JSModuleDef *quv__load_http(JSContext *ctx, const char *url) {
 JSModuleDef *quv_module_loader(JSContext *ctx, const char *module_name, void *opaque) {
     JSModuleDef *m;
     JSValue func_val;
+    int r;
+    DynBuf dbuf;
 
 #ifdef QUV_HAVE_CURL
     if (strncmp(http, module_name, strlen(http)) == 0 || strncmp(https, module_name, strlen(https)) == 0) {
         return quv__load_http(ctx, module_name);
     }
 #endif
 
+    dbuf_init(&dbuf);
+    r = quv__load_file(ctx, &dbuf, module_name);
+    if (r != 0) {
+        dbuf_free(&dbuf);
+        JS_ThrowReferenceError(ctx, "could not load '%s'", module_name);
+        return NULL;
+    }
+
     /* compile the module */
-    func_val = QUV_EvalFile(ctx, module_name, JS_EVAL_TYPE_MODULE | JS_EVAL_FLAG_COMPILE_ONLY);
+    func_val = JS_Eval(ctx, (char *) dbuf.buf, dbuf.size, module_name, JS_EVAL_TYPE_MODULE | JS_EVAL_FLAG_COMPILE_ONLY);
+    dbuf_free(&dbuf);
     if (JS_IsException(func_val)) {
         JS_FreeValue(ctx, func_val);
         return NULL;
     }
+
+    /* XXX: could propagate the exception */
+    js_module_set_import_meta(ctx, func_val, TRUE, FALSE);
     /* the module is already referenced, so we must free it */
     m = JS_VALUE_GET_PTR(func_val);
     JS_FreeValue(ctx, func_val);
 
     return m;
 }
+
+int js_module_set_import_meta(JSContext *ctx, JSValueConst func_val, JS_BOOL use_realpath, JS_BOOL is_main) {
+    JSModuleDef *m;
+    char buf[PATH_MAX + 16], *res;
+    JSValue meta_obj;
+    JSAtom module_name_atom;
+    const char *module_name;
+
+    CHECK_EQ(JS_VALUE_GET_TAG(func_val), JS_TAG_MODULE);
+    m = JS_VALUE_GET_PTR(func_val);
+
+    module_name_atom = JS_GetModuleName(ctx, m);
+    module_name = JS_AtomToCString(ctx, module_name_atom);
+    JS_FreeAtom(ctx, module_name_atom);
+    if (!module_name)
+        return -1;
+    if (!strchr(module_name, ':')) {
+        strcpy(buf, "file://");
```

Comment:
Suggestion: maybe use `pstrcpy()` instead? When I see `strcpy()`, I always have to do a double take to check whether the buffer is big enough to hold the input. In this case it does but it's not _immediately_ obvious.

Reply:
Good call. This one came straight from upstream, but yeah, makes sense!

Reply:
https://github.com/saghul/quv/pull/21